src/internal/stream.settings.cpp

```diff
@@ -0,0 +1,539 @@
+                          level);
+
+    settings->compression_level = level;
+
+    return ZarrError_Success;
+}
+
+ZarrError
+ZarrStreamSettings_set_compression_shuffle(ZarrStreamSettings* settings,
+                                           uint8_t shuffle)
+{
+    EXPECT_VALID_ARGUMENT(settings, "Null pointer: settings");
+    EXPECT_VALID_ARGUMENT(shuffle == BLOSC_NOSHUFFLE ||
+                            shuffle == BLOSC_SHUFFLE ||
+                            shuffle == BLOSC_BITSHUFFLE,
+                          "Invalid shuffle: %d. Must be %d (no shuffle), %d "
+                          "(byte shuffle), or %d (bit shuffle)",
+                          shuffle,
+                          BLOSC_NOSHUFFLE,
+                          BLOSC_SHUFFLE,
+                          BLOSC_BITSHUFFLE);
+
+    settings->compression_shuffle = shuffle;
+    return ZarrError_Success;
+}
+
+ZarrError
+ZarrStreamSettings_reserve_dimensions(ZarrStreamSettings* settings,
+                                      size_t count)
+{
+    EXPECT_VALID_ARGUMENT(settings, "Null pointer: settings");
+    EXPECT_VALID_ARGUMENT(count >= ZARR_DIMENSION_MIN &&
+                            count <= ZARR_DIMENSION_MAX,
+                          "Invalid count: %zu. Count must be between %d and %d",
+                          count,
+                          ZARR_DIMENSION_MIN,
+                          ZARR_DIMENSION_MAX);
+
+    settings->dimensions.resize(count);
+    return ZarrError_Success;
+}
+
+ZarrError
+ZarrStreamSettings_set_dimension(ZarrStreamSettings* settings,
+                                 size_t index,
+                                 const char* name,
+                                 size_t bytes_of_name,
+                                 ZarrDimensionType kind,
+                                 uint32_t array_size_px,
+                                 uint32_t chunk_size_px,
+                                 uint32_t shard_size_chunks)
+{
+    EXPECT_VALID_ARGUMENT(settings, "Null pointer: settings");
+    EXPECT_VALID_ARGUMENT(name, "Null pointer: name");
+    EXPECT_VALID_ARGUMENT(
+      bytes_of_name > 0, "Invalid name length: %zu", bytes_of_name);
+    EXPECT_VALID_ARGUMENT(strnlen(name, bytes_of_name) > 0,
+                          "Invalid name. Must not be empty");
+    EXPECT_VALID_ARGUMENT(
+      kind < ZarrDimensionTypeCount, "Invalid dimension type: %d", kind);
+    EXPECT_VALID_ARGUMENT(
+      chunk_size_px > 0, "Invalid chunk size: %zu", chunk_size_px);
+
+    // Check that the index is within bounds
+    if (index >= settings->dimensions.size()) {
+        LOG_ERROR("Invalid index: %zu. Must be less than %zu",
+                  index,
+                  settings->dimensions.size());
+        return ZarrError_InvalidIndex;
+    }
+
+    std::string dim_name = trim(name, bytes_of_name);
+    if (dim_name.empty()) {
+        LOG_ERROR("Invalid name. Must not be empty");
+        return ZarrError_InvalidArgument;
+    }
```

Comment:
- You already checked that `name` is not null, and that `bytes_of_name` is greater than zero. Under what circumstances would `dim_name` be empty?
- If such a case exists, it would be better to check it against a `std::string_view` to avoid a premature allocation of `std::string`.